### PR TITLE
USACO Plat friendcross solution

### PR DESCRIPTION
**Pull request by dongliu0426**

### CODE REVIEW

File: solutions/USACO feb17-friendcross.mdx [hunk 1]
```diff
@@ -396,4 +396,75 @@ int main() {
 
 ## Solution 3
 
-After generating the appropriate sequence of updates and queries, can apply divide & conquer as described in the module.
\ No newline at end of file
+After generating the appropriate sequence of updates and queries, can apply divide & conquer as described in the module.
+
+## Solution 4
+
+Let us consider an easier version: where we're just asked to count number of crossing pairs (ignoring whether the cows are friendly or not). To calculate the number of crossing pairs, we can add the cows in road 1 in reverse, while counting the number of cows that are added before in road 2. We can use a binary indexed tree to count this. This takes $\mathcal O(N)$ space and $\mathcal O(N\log N)$ time.
```
Comment (bqi343) [inline solutions/USACO feb17-friendcross.mdx:403]: > Let us consider an easier version: where we're just asked to count number of crossing pairs (ignoring whether the cows are friendly or not).

Let us consider an easier version; suppose that we just want to count the number of crossing pairs (ignoring whether the cows are friendly or not).

File: solutions/USACO feb17-friendcross.mdx [hunk 2]
```diff
@@ -396,4 +396,75 @@ int main() {
...
+
+To count the number of unfriendly crossings, instead of keeping track of the count for each BIT node, we keep track a list representing the cow breeds. Now to query the range for a breed $b$, we count the number of breeds in the range $[1, b-k-1]$ and $[b+k+1, n]$. We can use a [Order Statistic Tree
+](https://usaco.guide/gold/PURS?lang=cpp#order-statistic-tree) in C++ to represent the cow breeds. Since each update takes $\log N$ operations, it would allocate $\log N$ nodes, thus resulting with $\mathcal O(N\log N)$ space and $\mathcal O(N\log N)$ time.
```
Comment (bqi343) [inline solutions/USACO feb17-friendcross.mdx:406]: not $\mathcal O(N\log N)$ time

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by dongliu0426** [comment]
honestly idk if this is a good solution; it passes with `1957ms` on test case 13

File: solutions/USACO feb17-friendcross.mdx [hunk 3]
```diff
@@ -396,4 +396,87 @@ int main() {
...
 
-After generating the appropriate sequence of updates and queries, can apply divide & conquer as described in the module.
\ No newline at end of file
+After generating the appropriate sequence of updates and queries, can apply divide & conquer as described in the module.
+
+## Solution 4
+
+Let us consider an easier version; suppose that we just want to count the number of crossing pairs (ignoring whether the cows are friendly or not). To calculate the number of crossing pairs, we can add the cows in road 1 in reverse, while counting the number of cows that are added before in road 2. We can use a binary indexed tree to count this. This takes $\mathcal O(N)$ space and $\mathcal O(N\log N)$ time.
+
+To count the number of unfriendly crossings, instead of keeping track of the count for each BIT node, we keep track a list representing the cow breeds. Now to query the range for a breed $b$, we count the number of breeds in the range $[1, b-k-1]$ and $[b+k+1, n]$. We can use a [Order Statistic Tree
```
Comment (bqi343) [inline solutions/USACO feb17-friendcross.mdx:405]: > keep track a list

of a sorted set?

Comment (bqi343) [reply solutions/USACO feb17-friendcross.mdx:405]: > Now to query the range for a breed $b$, we count the num ...

For a breed $b$, we query the number of breeds ...

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by bqi343** [comment]
> honestly idk if this is a good solution; it passes with `1957ms` on test case 13

Then add a note.

File: solutions/USACO feb17-friendcross.mdx [hunk 4]
```diff
@@ -402,12 +402,14 @@ After generating the appropriate sequence of updates and queries, can apply divi
...
 
-To count the number of unfriendly crossings, instead of keeping track of the count for each BIT node, we keep track a list representing the cow breeds. Now to query the range for a breed $b$, we count the number of breeds in the range $[1, b-k-1]$ and $[b+k+1, n]$. We can use a [Order Statistic Tree
-](https://usaco.guide/gold/PURS?lang=cpp#order-statistic-tree) in C++ to represent the cow breeds. Since each update takes $\log N$ operations, it would allocate $\log N$ nodes, thus resulting with $\mathcal O(N\log N)$ space and $\mathcal O(N\log^2 N)$ time.
+To count the number of unfriendly crossings, instead of keeping track of the count for each BIT node, we keep track a sorted set representing the cow breeds. For a breed $b$, we query the number of breeds in the range $[1, b-k-1]$ and $[b+k+1, n]$. We can use a [Order Statistic Tree
```
Comment (bqi343) [inline solutions/USACO feb17-friendcross.mdx:405]: keep track **of** a sorted set (or "maintain a sorted set")

File: solutions/USACO feb17-friendcross.mdx [hunk 5]
```diff
@@ -402,12 +402,14 @@ After generating the appropriate sequence of updates and queries, can apply divi
...
+](https://usaco.guide/gold/PURS?lang=cpp#order-statistic-tree) in C++ to represent the cow breeds. Since each update takes $\log N$ operations, it would allocate $\log N$ nodes each update, thus resulting with $\mathcal O(N\log N)$ space and $\mathcal O(N\log^2 N)$ time.
```
Comment (bqi343) [inline solutions/USACO feb17-friendcross.mdx:406]: > thus resulting with 

resulting in?

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Pull request merged**